harness/api_checkers.go

```diff
@@ -0,0 +1,493 @@
+type APICheckSuite struct {
+	suite.Suite
+	opts      APICheckCapabilities
+	newTracer func() (tracer opentracing.Tracer, closer func())
+	tracer    opentracing.Tracer
+	closer    func()
+}
+
+// RunAPIChecks runs a test suite to check a Tracer against the OpenTracing API.
+// It is provided a function that will be executed to create and destroy a tracer for each test
+// in the suite, and the given APICheckOption functional options `opts`.
+func RunAPIChecks(
+	t *testing.T,
+	newTracer func() (tracer opentracing.Tracer, closer func()),
+	opts ...APICheckOption,
+) {
+	s := &APICheckSuite{newTracer: newTracer}
+	for _, opt := range opts {
+		opt(s)
+	}
+	suite.Run(t, s)
+}
+
+// APICheckOption instances may be passed to NewAPICheckSuite.
+type APICheckOption func(*APICheckSuite)
+
+// CheckBaggageValues returns an option that sets whether to check for propagation of baggage values.
+func CheckBaggageValues(val bool) APICheckOption {
+	return func(s *APICheckSuite) {
+		s.opts.CheckBaggageValues = val
+	}
+}
+
+// CheckExtract returns an option that sets whether to check if extracting contexts from carriers works.
+func CheckExtract(val bool) APICheckOption {
+	return func(s *APICheckSuite) {
+		s.opts.CheckExtract = val
+	}
+}
+
+// CheckInject returns an option that sets whether to check if injecting contexts works.
+func CheckInject(val bool) APICheckOption {
+	return func(s *APICheckSuite) {
+		s.opts.CheckInject = val
+	}
+}
+
+// CheckEverything returns an option that enables all API checks.
+func CheckEverything() APICheckOption {
+	return func(s *APICheckSuite) {
+		s.opts.CheckBaggageValues = true
+		s.opts.CheckExtract = true
+		s.opts.CheckInject = true
+	}
+}
+
+// UseProbe returns an option that specifies an APICheckProbe implementation to use.
+func UseProbe(probe APICheckProbe) APICheckOption {
+	return func(s *APICheckSuite) {
+		s.opts.Probe = probe
+	}
+}
+
+// BeforeTest creates a tracer for this specific test invocation.
+func (s *APICheckSuite) BeforeTest(suiteName, testName string) {
+	s.tracer, s.closer = s.newTracer()
+	if s.tracer == nil {
+		panic(fmt.Sprintf("newTracer returned nil Tracer before running %s, %s", suiteName, testName))
+	}
+}
+
+// AfterTest closes the tracer, and clears the test-specific tracer.
+func (s *APICheckSuite) AfterTest(suiteName, testName string) {
+	if s.closer != nil {
+		s.closer()
+	}
+	s.tracer, s.closer = nil, nil
+}
+
+// TestStartSpan checks if a Tracer can start a span and calls some span API methods.
+func (s *APICheckSuite) TestStartSpan() {
+	span := s.tracer.StartSpan(
+		"Fry",
+		opentracing.Tag{Key: "birthday", Value: "August 14 1974"})
+	span.LogFields(
+		log.String("hospital", "Brooklyn Pre-Med Hospital"),
+		log.String("city", "Old New York"))
+	span.Finish()
+}
+
+// TestStartSpanWithParent checks if a Tracer can start a span with a specified parent.
+func (s *APICheckSuite) TestStartSpanWithParent() {
+	parentSpan := s.tracer.StartSpan("parent")
```

Comment:
s/"parent"/"Turanga Morris"/  (or Turanga Munda). 

Damn, we need a ref type for >1 parent! :-)

Reply:
Sure, I agree!